lib/apiservers/engine/backends/filter/container.go

```diff
@@ -0,0 +1,228 @@
+// Copyright 2017 VMware, Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package filter
+
+import (
+	"fmt"
+	"strconv"
+
+	"github.com/docker/engine-api/types"
+
+	"github.com/vmware/vic/lib/apiservers/engine/backends/cache"
+	"github.com/vmware/vic/lib/apiservers/portlayer/models"
+)
+
+// reused from docker/docker/daemon/list.go
+type ContainerListContext struct {
+	FilterContext
+
+	// Counter is the container iteration index for this context
+	Counter int
+	// ExitCode for the passed container
+	ExitCode int
+	// exitAllowed is a list of exit codes allowed to filter with
+	exitAllowed map[int]struct{}
+	// ContainerListOptions is the filters set by the user
+	*types.ContainerListOptions
+}
+
+/*
+*	IncludeContainer will evaluate the filter criteria in listContext against the provided
+* 	container and determine what action to take.  There are three options:
+*		* IncludeAction
+*		* ExcludeAction
+*		* StopAction
+*
+ */
+func IncludeContainer(listContext *ContainerListContext, container *models.ContainerInfo) FilterAction {
+
+	// if we need to filter on name add to the listContext
+	if listContext.Filter.Include("name") {
+		// containerConfig allows for multiple names, but only 1 ever
+		// assigned
+		listContext.Name = container.ContainerConfig.Names[0]
+	}
+
+	// filter common requirements
+	act := filterCommon(&listContext.FilterContext, listContext.Filter)
+	if act != IncludeAction {
+		return act
+	}
+
+	// Stop iteration when the index is over the limit
+	if listContext.Limit > 0 && listContext.Counter == listContext.Limit {
+		return StopAction
+	}
+
+	// Do not include container if its exit code is not in the filter
+	if len(listContext.exitAllowed) > 0 {
```

Comment:
I don't think the map adds clarity as it stands; I had expected the entire clause to be simplified if switching to a map:
```go
_, ok := listContext.exitAllowed[listContext.ExitCode]
if !ok || container.ProcessConfig.ExitCode != listContext.ExitCode || container.container.ContainerConfig.State == "Running" || container.ProcessConfig.StartTime == 0 {
	return ExcludeAction
}
```

If keeping it as an array, reversing the order of the tests avoids the iteration if not necessary:
```go
if len(listContext.exitAllowed) > 0 {
   // is container in valid state for exitcode check
    if container.ContainerConfig.State == "Running" || container.ProcessConfig.StartTime == 0 {
	return ExcludeAction
    }

    ok := false
    for _, code := range exitAllowed {
        if code == listContext.ExitCode {
            ok = true
            break
        }
    }
    if !ok {
        return ExcludeAction
    }
}
```